File: modules/swagger-core/src/test/scala/converter/JodaLocalDateConverterTest.scala
package converter

import com.wordnik.swagger.annotations.ApiModelProperty
import com.wordnik.swagger.converter.ModelConverters
import com.wordnik.swagger.core.util.JsonSerializer
import org.joda.time.LocalDate
import org.junit.runner.RunWith
import org.scalatest.FlatSpec
import org.scalatest.junit.JUnitRunner
import org.scalatest.matchers.ShouldMatchers

import scala.annotation.meta.field

@RunWith(classOf[JUnitRunner])
class JodaLocalDateConverterTest extends FlatSpec with ShouldMatchers {
  it should "read a generic model" in {
    val models = ModelConverters.readAll(classOf[ModelWithJodaLocalDate])
    models.size should be(1)

    val model = models.head
    val nameProperty = model.properties("name")
    nameProperty.`type` should be("string")
    nameProperty.position should be(2)
    nameProperty.description should be(Some("name of the model"))

    val dateTimeProperty = model.properties("createdAt")
    dateTimeProperty.`type` should be("Date")
    dateTimeProperty.position should be(1)
    dateTimeProperty.required should be(true)
    dateTimeProperty.description should be(Some("reation localDate"))

    println(JsonSerializer.asJson(models.head))
  }
}


case class ModelWithJodaLocalDate(
  @(ApiModelProperty@field)(value = "name of the model", position = 2) name: String,
  @(ApiModelProperty@field)(value = "creation localDate", required = true, position = 1) createdAt: LocalDate)

File: modules/swagger-core/src/test/java/io/swagger/v3/core/resolving/JodaLocalDateConverterTest.java
/**
 * Copyright 2021 SmartBear Software
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.swagger.v3.core.resolving;

import io.swagger.v3.core.converter.ModelConverters;
import io.swagger.v3.oas.models.media.DateSchema;
import io.swagger.v3.oas.models.media.DateTimeSchema;
import io.swagger.v3.oas.models.media.Schema;
import io.swagger.v3.oas.models.media.StringSchema;
import org.joda.time.LocalDate;
import org.testng.annotations.Test;

import java.time.Instant;
import java.util.Map;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertTrue;

public class JodaLocalDateConverterTest {

    @Test
    public void testJodaLocalDate() {
        final Map<String, Schema> models = ModelConverters.getInstance().read(ModelWithJodaLocalDate.class);
        assertEquals(models.size(), 1);

        final Schema model = models.get("ModelWithJodaLocalDate");

        final Schema dateTimeProperty = (Schema) model.getProperties().get("createdAt");
        assertTrue(dateTimeProperty instanceof DateSchema);
        assertTrue(model.getRequired().contains("createdAt"));
        assertEquals(dateTimeProperty.getDescription(), "creation localDate");

        final Schema nameProperty = (Schema) model.getProperties().get("name");
        assertTrue(nameProperty instanceof StringSchema);
        assertEquals(nameProperty.getDescription(), "name of the model");
    }

    class ModelWithJodaLocalDate {
        @io.swagger.v3.oas.annotations.media.Schema(description = "name of the model")
        public String name;

        @io.swagger.v3.oas.annotations.media.Schema(description = "creation localDate", required = true)
        public LocalDate createdAt;
    }

    @Test
    public void testJavaTimeInstant() {
        final Map<String, Schema> models = ModelConverters.getInstance().read(ModelWithJavaTimeInstant.class);
        assertEquals(models.size(), 1);
        final Schema model = models.get("ModelWithJavaTimeInstant");

        final Schema dateTimeProperty = (Schema) model.getProperties().get("createdAt");
        assertTrue(dateTimeProperty instanceof DateTimeSchema);
    }

    class ModelWithJavaTimeInstant {
        @io.swagger.v3.oas.annotations.media.Schema(description = "name of the model")
        public String name;
        public Instant createdAt;
    }
}
